File: references/betterlectio/android/app/src/main/java/dk/betterlectio/android/feature/referral/ReferralCoordinator.kt
package dk.betterlectio.android.feature.referral

import com.posthog.PostHog
import dk.betterlectio.android.core.model.Student
import dk.betterlectio.android.feature.supabase.SupabaseReferralService
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import timber.log.Timber
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Orchestrates Install Referrer → referral-finalize and soft share nudges.
 */
@Singleton
class ReferralCoordinator @Inject constructor(
    private val referrerReader: InstallReferrerReader,
    private val referralService: SupabaseReferralService,
    private val store: ReferralStore,
) {
    private val _nudgeVisible = MutableStateFlow(false)
    val nudgeVisible: StateFlow<Boolean> = _nudgeVisible.asStateFlow()

    private val _celebrationName = MutableStateFlow<String?>(null)
    val celebrationName: StateFlow<String?> = _celebrationName.asStateFlow()

    private val _cachedStats = MutableStateFlow<ReferralStats?>(null)
    val cachedStats: StateFlow<ReferralStats?> = _cachedStats.asStateFlow()

    /**
     * Call after Supabase session is ready (login or cold start). At-most-once per student.
     */
    suspend fun tryFinalizeAfterAuth(student: Student) {
        if (student.isDemo) return
        if (store.wasFinalizeAttempted(student.studentId)) return

        val cookieId = referrerReader.readReferralCookieId()
        // Mark attempted after we got a referrer response (even empty) so we don't
        // hammer Play's API — but only after a real network finalize attempt when
        // we have a cookie, or immediately when there is nothing to attribute.
        if (cookieId.isNullOrBlank()) {
            store.markFinalizeAttempted(student.studentId)
            Timber.d("Referral: no Install Referrer cookie — skipping finalize")
            return
        }

        val result = referralService.finalizeAndroid(
            studentId = student.studentId,
            schoolId = student.gymId,
            cookieId = cookieId,
        )
        // Only mark attempted on a definite HTTP response (success or attributed:false).
        // Network failures leave the flag unset so cold-start can retry.
        if (result != null) {
            store.markFinalizeAttempted(student.studentId)
            if (result.attributed) {
                PostHog.capture(
                    event = "referral attributed",
                    properties = mapOf(
                        "platform" to "android",
                        "referrer_student_id" to (result.referrerStudentId ?: ""),
                    ),
                )
            }
            Timber.i(
                "Referral finalize: attributed=%s reason=%s",
                result.attributed,
                result.reason,
            )
        }
    }

    suspend fun refreshStats(studentId: String): ReferralStats? {
        val stats = referralService.getStats(studentId) ?: return _cachedStats.value
        val previous = store.lastKnownConversions(studentId)
        if (previous >= 0 && stats.conversions > previous) {
            val newest = stats.recentReferrals.firstOrNull()?.name
            _celebrationName.value = newest ?: "En ven"
        }
        store.setLastKnownConversions(studentId, stats.conversions)
        _cachedStats.value = stats
        return stats
    }

    fun consumeCelebration() {
        _celebrationName.value = null
    }

    /**
     * After first successful skema load — show soft nudge once if under unlock threshold.
     */
    suspend fun maybeShowNudge(student: Student) {
        if (student.isDemo) return
        if (store.wasNudgeShown(student.studentId)) return
        val stats = refreshStats(student.studentId) ?: return
        if (referralUnlockProgress(stats.conversions).unlocked) {
            store.markNudgeShown(student.studentId)
            return
        }
        _nudgeVisible.value = true
    }

    fun dismissNudge(studentId: String) {
        store.markNudgeShown(studentId)
        _nudgeVisible.value = false
    }
}

File: references/betterlectio/ios/BetterLectio/ReferralCoordinator.swift
import Combine
import Foundation

@MainActor
final class ReferralCoordinator: ObservableObject {
    static let shared = ReferralCoordinator()

    @Published private(set) var cachedStats: ReferralStats?
    @Published var nudgeVisible = false
    @Published var celebrationName: String?
    @Published private(set) var lastAttribution: ReferralFinalizeResult?
    private var activeStudentID: String?
    private var captureInFlight = false

    private let service: SupabaseReferralService
    private let store: ReferralStore

    init(service: SupabaseReferralService = .shared, store: ReferralStore = .shared) {
        self.service = service
        self.store = store
    }

    func cachedStats(for student: Student) -> ReferralStats? {
        activeStudentID == student.studentId ? cachedStats : nil
    }

    func activate(for student: Student) {
        guard activeStudentID != student.studentId else { return }
        resetSession()
        activeStudentID = student.studentId
    }

    func resetSession() {
        activeStudentID = nil
        cachedStats = nil
        nudgeVisible = false
        celebrationName = nil
        lastAttribution = nil
    }

    func handle(url: URL, authenticatedStudent: Student?) async {
        guard let link = ReferralLink.parse(url) else { return }
        guard !captureInFlight else { return }
        captureInFlight = true
        defer { captureInFlight = false }
        do {
            let token: UUID
            if let candidate = link.token {
                guard try await service.validate(token: candidate, referrerStudentID: link.studentID) else { return }
                token = candidate
            } else {
                token = try await service.registerClick(referrerStudentID: link.studentID)
            }
            _ = store.saveFirstPending(token: token)
            if let authenticatedStudent {
                await finalizeIfNeeded(for: authenticatedStudent)
            }
        } catch {
            print("⚠️ [Referral] Could not capture link: \(error.localizedDescription)")
        }
    }

    func finalizeIfNeeded(for student: Student) async {
        activate(for: student)
        guard !student.isDemo,
              !store.wasFinalizeAttempted(studentID: student.studentId),
              let pending = store.pending() else { return }
        do {
            let result = try await service.finalize(student: student, token: pending.token)
            lastAttribution = result
            store.markFinalizeAttempted(studentID: student.studentId)
            store.clearPending()
        } catch {
            // A transport/server failure is intentionally retryable on next launch.
            print("⚠️ [Referral] Finalization deferred: \(error.localizedDescription)")
        }
    }

    @discardableResult
    func refreshStats(for student: Student) async -> ReferralStats? {
        activate(for: student)
        if student.isDemo {
            cachedStats = .demo
            return .demo
        }
        do {
            let stats = try await service.stats(studentID: student.studentId)
            if let previous = store.lastKnownConversions(studentID: student.studentId),
               stats.conversions > previous {
                celebrationName = stats.recentReferrals.first?.name ?? "En klassekammerat"
            }
            store.setLastKnownConversions(stats.conversions, studentID: student.studentId)
            cachedStats = stats
            return stats
        } catch {
            print("⚠️ [Referral] Stats unavailable: \(error.localizedDescription)")
            return cachedStats(for: student)
        }
    }

    func maybeShowNudge(for student: Student) async {
        guard !student.isDemo, !store.wasNudgeShown(studentID: student.studentId) else { return }
        guard let stats = await refreshStats(for: student) else { return }
        if ReferralProgress(conversions: stats.conversions).unlocked {
            store.markNudgeShown(studentID: student.studentId)
        } else {
            nudgeVisible = true
        }
    }

    func dismissNudge(for student: Student) {
        store.markNudgeShown(studentID: student.studentId)
        nudgeVisible = false
    }
}
